### PR TITLE
[WIP] New Zig formal grammar

### PR DESCRIPTION
**Pull request by Hejsil**
This is an attempt at formalizing a Parsing Expression Grammar for the Zig programming language. This is done to find a better solution for #760.

Currently, I have the grammar posted [here](https://github.com/Hejsil/zig-grammar2) using the  [peg](http://piumarta.com/software/peg/) parser generator to validate it. The grammar is a breaking change from 0.3.0 (See what changed in db5d479).

The grammar implements #1047 and some of #114.

### CODE REVIEW

**Comment by andrewrk** [comment]
Thanks for doing this work.

> (See what changed in [0de46ff](https://github.com/ziglang/zig/commit/0de46ff507739664ba1f20d17f4e503400bd9312)).

```zig
try (switch (c) {
```

This seems worse. Why is this necessary?

```zig
         if (base.id == (comptime typeToId(T))) { 
```

Same question

```zig
     link_err: errorset{OutOfMemory}!void, 
```

Can we keep `error` as the keyword that makes error sets, and use `anyerror` as the new global error set primitive type?

```zig
return HashInt(unsigned_x) ^ (comptime rng.scalar(HashInt));
```

Would this work? `return HashInt(unsigned_x) ^ comptime(rng.scalar(HashInt));`

```zig
pub const LPOVERLAPPED_COMPLETION_ROUTINE = ?(extern fn (DWORD, DWORD, *OVERLAPPED) void);
```

This seems worse. Why is this needed?

```zig
    assert(1234 == (switch (x) {
        MultipleChoice.A => 1,
        MultipleChoice.B => 2,
        MultipleChoice.C => u32(1234),
        MultipleChoice.D => 4,
    })); 
```

Same question.

```zig
    if (t or (x: {
        assert(f);
        break :x f;
    })) {
```

Same question.



> The 1 token lookahead goal might not be an interesting goal to reach, but we are really close

I agree that 1 token lookahead is not an important goal to reach. I will happily trade a couple token lookaheads for any other syntactic gain.

> It would also nice we had the bison parser the compiled on every commit since bison can detect ambiguaties through conflicts. Using the parser on all `.zig` files would also be a good idea, so that we ensure that stage1 and 2 conforms to the spec.

Let's keep the dependencies of Zig at a minimum, that is, a system c++ compiler, libLLVM, and libclang. However I would be open to a separate repository dedicated to testing Zig grammar, which we could have the CI use to run on every commit.

**Comment by Hejsil** [comment]
The reason switch and comptime require parens are because I gave them a high precedence, together with the other control flow expressions.

For comptime, it was done because of this mess:
```rust
// A lower percedence `comptime` Expr rule would cause this ambiguity:
async<comptime A> fn()void
//async<(comptime A> fn()void)
//async<(comptime A)> fn()void
```

I think we could also solve this with `comptime (expr)`.

For switch and blocks, this was done to correctly formalise the rule that these statements shouldn't have semicolons behind them:
```rust
// Prev rules:
// Statement
//    : SwitchExpr
//    | Expr Semicolon
//    ...
// PrimaryExpr
//    : SwitchExpr
//    ...
// All these are valid, with the former grammar
switch (a) {}
switch (a) {};
{}
{};

// It would also cause this ambiguity
{}{}; // Is this two blocks, or a block followed by an initializer
switch (a) {}{}; // Same with switch
```

The requirement of parens around fn types is to resolve this:
```rust
fn()fn()void!void
//fn()(fn()void!void) // This is how it is parsed with the new grammar
//fn()(fn()void)!void

// This solution requires this grammar
// FnTypeExpr
//     : ErrorUnionExpr
//     | FnTypePrefix ErrorUnionExpr
// 
// ErrorUnionExpr
//     : PrefixExpr
//     | PrefixExpr ExclamationMark PrefixExpr
// 
// PrefixExpr
//     : SuffixExpr
//     | PrefixOp PrefixExpr // To have []fn()void, parens is required
```

I'll look more into laxing these paren requirements. If you have any idea's, I'm all ears :)

**Comment by Hejsil** [comment]
I like the seperate repo idea btw. Where do we keep the grammar? In the Zig or Zig-grammar repo?

**Comment by Hejsil** [comment]
>Can we keep error as the keyword that makes error sets, and use anyerror as the new global error set primitive type?

We could, but what about`error.SomeName`.

**Comment by andrewrk** [comment]
> Where do we keep the grammar?

In the separate repo I think. ziglang/zig is an implementation of the zig specification (which isn't written yet; see #75) using recursive descent, and the grammar repo would be a tool used for validating and testing the formal grammar specification.

> We could, but what about`error.SomeName`.

We could make that continue to work with special syntax, yes? It's always been syntactic sugar for `error{SomeName}.SomeName`.

**Comment by Hejsil** [comment]
>We could make that continue to work with special syntax, yes? It's always been syntactic sugar for error{SomeName}.SomeName.

Right, we could do that. Was trying to keep the `Expr Dot Symbol` to also handle this case, but I guess there is really no need for that. We can just have this `PrimaryExpr: Keyword_error Dot Identifier | ...`

**Comment by winksaville** [comment]
Since were formalizing the grammer I'd like to suggest allowing seperators in numeric literals to improve readability.

I did some research and [C++14](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3781.pdf) uses single quote `'` and other languages use underscore `_`; [rust](https://doc.rust-lang.org/rust-by-example/primitives/literals.html), [java](https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html), [swift](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html), [ruby](https://docs.ruby-lang.org/en/2.0.0/syntax/literals_rdoc.html#label-Numbers), [perl](https://www.linuxtopia.org/Perl_Programming/pickingUpPerl_25.html), [python](https://www.python.org/dev/peps/pep-0515/) and maybe others.

Of course there is at least one language where it was discussed and rejected, Go [here](https://groups.google.com/forum/#!msg/Golang-nuts/nVsE8EL39jg/WtwY6UBzAQAJ) and [here](https://github.com/golang/go/issues/42).

**Comment by UniqueID1** [comment]
> There are still a few ambiguous that I'm not sure how to fix:
> 
> ```
> // async<A> (fn()void) ()
> // (async<A> fn()void) ()
> _ = async<A> fn()void ();
> ```

'>' acts exactly like '{' in this case. In some places an expression can not contain '{' it always starts the function body. Here '>' always closes async. In both cases you can allow the use of parentheses to have the code parsed differently. You could always unify these simpler expressions, disallowing both '{' and '>' in both cases.

The same could be said about '[' and ']'. When inside '[' the next ']' does not apply to the current expression but to the parent. Ignoring the fact that Zig does not have a ']' operator, but that's besides the point.
```
// A ']' operator would work fine
const a = b ] c;
const z = b[(c]d)];
```
`async<comptime A> fn()void` also works fine.

You can also choose to favor the grammar and parser writer. But it's not because a context free grammar can't be created that handles what Zig mostly looks like today.

**Comment by UniqueID1** [comment]
From the compiler writer's POV, I think it's really just about keeping track of what token stops the current expression and returns to the parent. Sometimes it's ']', sometimes it's ',', sometimes it's '{', sometimes it's '>'.

**Comment by Hejsil** [comment]
> > There are still a few ambiguous that I'm not sure how to fix:
> > ```
> > // async<A> (fn()void) ()
> > // (async<A> fn()void) ()
> > _ = async<A> fn()void ();
> > ```
> 
> '>' acts exactly like '{' in this case. In some places an expression can not contain '{' it always starts the function body. Here '>' always closes async. In both cases you can allow the use of parentheses to have the code parsed differently. You could always unify these simpler expressions, disallowing both '{' and '>' in both cases.

This is not a problem with `<>`. The problem still exists with this example:
 ```
 // async (fn()void) ()
 // (async fn()void) ()
 _ = async fn()void ();
 ```

The grammar is ambiguous between an async call and an async function type.

**Comment by UniqueID1** [comment]
Yes, my bad I did edit my answer.

**Comment by Hejsil** [comment]
@winksaville #504

**Comment by UniqueID1** [comment]
I think this is a nice attempt at making the grammar context free but now that the grammar is simpler for machines it needs to be refined for humans. These things stick out to me as being very awkward:

```
// unintuitive parentheses around if expr
const A = B + (if (rem == 0) 0 else (os.page_size - rem));

// unintuitive parentheses around return expr, now no parentheses around if expr
const A = if (B >= T.bit_count) (return 0) else @intCast(Log2Int(T), abs_shift_amt);

// no parentheses around second return stmt, parentheses around first
if (b) |b_p| (return eql(a_p, b_p)) else |_| return false;

// parentheses around if stmt
pub const ChildProcess = struct {
    pub pid: (if (is_windows) void else i32),
}
```

I shortened some of the identifier names.

**Comment by Hejsil** [comment]
@UniqueID1
I agree that the parens around the return expressions are bad, though I think some enforcement of parens around `if` is not too bad of an idea (maybe not when it's alone, but for some expressions, it would definitely improve readability).

**Comment by UniqueID1** [comment]
Let's clarify something (that I'm just learning). Even if we never get to a point where we have a BNF grammar that Bison parses without finding conflicts, the **grammar is still entirely context free** and significantly better and easier to parse than C/C++. We also have a parser in the std library.

Furthermore, according to wikipedia (on the dangling else problem): 

> If the parser is hand written, the programmer may use a non-ambiguous context-free grammar. Alternatively, one may rely on a non-context-free grammar or a parsing expression grammar.

So a grammar that's ambiguous when written in BNF can be thought of as being **unambiguous** when the parser is hand written.

Furthermore, an ambiguity issue such as a dangling else can be rewritten to avoid the conflict in LR parsers:

```
%%
statement: open_statement
         | closed_statement
         ;

expression: "a"
		  | %empty
		  ;

open_statement: "IF" '(' expression ')' simple_statement
              | "IF" '(' expression ')' open_statement
              | "IF" '(' expression ')' closed_statement "ELSE" open_statement
              ;

closed_statement: simple_statement
                | "IF" '(' expression ')' closed_statement "ELSE" closed_statement
                ;

simple_statement: "..."
                ;

%%

```

That's a "dangling else ambiguity" that all of a sudden is no more, as the else is always grouped with the nearest if. This is now **unambiguous**. This whole thing is I think just about shitty grammar 'languages'/theory stuff that needs to be bruteforced into understanding simple precedence. And the 'grammar' is really ugly as a result.

I hope I'm understanding this correctly. I wish we had people that were truly knowledgeable in this.

I think I know how to solve if (expr) if (expr) if (expr) return x; I've converted part of Zig's grammar to BNF and I'm going to be playing with it, huge thanks to Hejsil for the code that I've been learning from.

**Comment by wirelyre** [comment]
@UniqueID1 A *language* is a set of strings. In this case, the set of all syntactically legal Zig programs. A *grammar* is a structured way of representing a language.

One property of a language (but not a grammar) is whether it is context free. If you can write a grammar for Bison, then the language it parses is context free.

BNF grammars define languages unambiguously, but not *parse trees*. Bison emits a warning when there are two different ways to interpret the same input. This does not affect *which* inputs are accepted, but rather *why* they are accepted (that is, not the question "is this a legal program?", but rather "what is the structure of this program?").

It is not really enough to write a grammar, but let the resolution of ambiguous parses depend on a hand-written parser. Now you have to understand the parser program; and additionally, if the parser program does not correctly implement the language defined by the grammar, you're sunk.

**Comment by UniqueID1** [comment]
@wirelyre Thanks. I deleted my comment.

**Comment by Hejsil** [comment]
I've thought long and hard about this, and I don't think we can make this grammar be context-free and unambiguous without sacrificing some of the niceness of the current syntactic constructs, or resorting to a special priority system outside the grammar (which kinda defeats the point).

I propose that we instead create the grammar as a [Parsing expression grammar](https://en.wikipedia.org/wiki/Parsing_expression_grammar). The pros here are, that the grammar cannot be ambiguous as only the first matching rule will be considered. The cons to this approach are, that PEGs hide syntax flaws.

Here is how I imagine all the current ambiguities will be parsed if we have a PEG:
```

/  FnDef   \
fn a() !b {}
       ^|
ErrorInf|
     TypeExpr

/          IfExpr          \
          /     IfExpr     \
if (true) if (true) 1 else 1

/   Async call    \
         /FnType\
async<A> fn()void()

/    Call    \
/Async call\
async<A> a()()

/   FnType    \
    /  FnType \
        /ErrUn\
fn()fn()void!u8

/         FnType         \
      /TypeExpr\
async<comptime A> fn()void;

// This one will be hard to parse correctly, since, whith a recursive decent parser
// we will start by parsing it as:
      /    TypeExpr      \
async<comptime A> fn()void;

// The parser should then revert its current parsing, and parse it correctly...
// I think we'll just parse this wrong until we have
// https://github.com/ziglang/zig/issues/1639
```

Also, we can't have the semicolon rule with PEG if we wanna keep the block expression having high priority:
```
if (true) {} // Valid
if (true) {}; // Valid, should be an error, but we can't have this without diallowing if {} in if expressions.
if (true) A;  // Valid
if (true) A  // error expected ;
```

**Comment by thejoshwolfe** [comment]
It seems like getting rid of <> as grouping operators would solve some problems. C++ has some nasty syntax due to `>` being both an infix operator and a group closing operator. Zig has the same problem with `async`.

I know the semantics of async are planned to change in the near future. Perhaps those changes will yield different syntax. I've always been uneasy with <> as grouping operators, so avoiding them entirely seems promising.

**Comment by andrewrk** [comment]
> I've always been uneasy with <> as grouping operators, so avoiding them entirely seems promising.

In #661 we need a grouping operator to pass a calling convention expression to the `fn` keyword. Do you have a syntax suggestion for this?

**Comment by andrewrk** [comment]
> I've thought long and hard about this, and I don't think we can make this grammar be context-free and unambiguous without sacrificing some of the niceness of the current syntactic constructs, or resorting to a special priority system outside the grammar (which kinda defeats the point).
> 
> I propose that we instead create the grammar as a [Parsing expression grammar](https://en.wikipedia.org/wiki/Parsing_expression_grammar). The pros here are, that the grammar cannot be ambiguous as only the first matching rule will be considered. The cons to this approach are, that PEGs hide syntax flaws.

I think this is the way to move forward. :+1: from me. This is always how I imagined the grammar working. I believe that I have incorrectly been using the term "context-free grammar" when I only meant "you can make a parse tree without doing any semantic analysis".

**Comment by Hejsil** [comment]
The prefix `async` to async calls is really never gonna work.
```
Option 1:
SuffixExpr
    <- AsyncPrefix SuffixExpr FnCallArgumnets
     / PrimaryExpr SuffixOp*

const a = async t.t();
AsyncPrefix: "async" Ok
SuffixExpr: "t.t()" Ok
FnCallArgumnets: ";" failed
PrimaryExpr: "async" failed

Option 2:
SuffixExpr <- AsyncCallExpr SuffixOp*
AsyncCallExpr
    <- AsyncPrefix PrimaryExpr FnCallArgumnets
     / PrimaryExpr

const a = async t.t();
AsyncPrefix: "async" Ok
PrimaryExpr: "t" Ok
FnCallArgumnets: "." failed
PrimaryExpr: "async" failed
```

As we can see, both of these examples cannot be parsed without the parser having to be context aware.

**Comment by andrewrk** [comment]
Can you elaborate a little? I'm focused on this copy elision stuff and so I'm not quite grokking your examples. In the current parser we look for a function call expression directly after `async` and it seems to work. Is there a problem with it?

**Comment by Hejsil** [comment]
@andrewrk What we do in both parsers, is look at what type the resulting expression of `SuffixExpr` is. If it's a call, then we add the async info to this node. This is parsing based on context.

**Comment by Hejsil** [comment]
What I show in my examples, is how the PEG grammar would parse the example. Failing, because we cannot express this check in the grammar.

**Comment by Hejsil** [comment]
This is the closest we will get:
```
SuffixExpr
    <- AsyncPrefix PrimaryExpr SuffixOp* FnCallArgumnets
     / PrimaryExpr (SuffixOp / FnCallArgumnets)*
```

But this will not allow:
```
// a is a function that returns an async function
async a()();
```

**Comment by andrewrk** [comment]
Ah, ok this is clear to me now. Thanks. OK let's assume a goal of making the async syntax work with PEG grammer. Backing up, here are the semantic needs (after the coroutine rework is done):

 * Some functions have async calling convention. This is important, and makes sense, because async functions are called differently. They have a secret first argument which is their frame, used instead of the stack.
 * Async functions are allowed to suspend. (Note that `await` is a form of suspending.) It is entirely possible that async functions could call other async functions without any special syntax. This would be equivalent to directly using `await` after doing an async call. However, this is undesirable for 2 reasons: 1. It is often useful for parallelism to async call, which gives a "promise" handle, and then later `await`. 2. Suspend points are semantically interesting and important, and it's best to not hide this control flow.
 * The key insight of the coroutines rewrite, is that with well defined copy elision, a simple assignment is the best way to initialize an async call: `frame = @asyncCallSyntax(async_function, ...);`. This is a function call that, under the hood, calls `async_function` by passing a secret first parameter which is the pointer to `frame` (which we have because it is the result location according to well defined copy elision). Here I used example syntax as a placeholder for whatever we come up with.
 * Next we need a way to `await` the frame, which can probably just remain `await frame`. The only other possible win here is that we could have a way to assert that the function has completed and obtain the return value without suspending. This would be useful when calling `asyncMain` from `main` for example. After `event_loop.run()` returns, we are guaranteed that the async function has completed and the return value is available.
 * We also need to detect when one wants to do `await` and `async` at the same time. There are some important optimizations and better codegen possible for this case. This enables the whole thing to be a tail call, which provides the solution to recursion using coroutines. If you had the frame memory on the heap, you would need to be able to pass it to the "await and async at the same time" syntax. When doing this kind of await/async call, this enables normal copy elision rules, whereas without it, coroutines involve copying the return value and (directly passed) parameters.
    * The current idea is that `await async` would be specially detected as this. However this does not solve passing in a frame that is heap allocated.

Can we come up with a syntax that will provide these semantics?

**Comment by Hejsil** [comment]
@andrewrk I'll think about this for a bit. On another note, I now have a PEG that can parse all the Zig code in this repo (using [this](https://github.com/ziglang/zig/pull/1685#issuecomment-434447473) hack for async calls). Changes from 0.3.0 are in db5d479 (error -> anyerror).

**Comment by wirelyre** [comment]
@Hejsil I would not be so hasty to throw out the current syntax entirely. It seems like the `async` keyword currently binds more weakly than `.` but equally strongly as the following `()`, right?

Maybe it makes sense to consider a "path" `root.branch.leaf` as a single lexical unit.

**Comment by Hejsil** [comment]
@wirelyre The problem here is, that what we really want, is for `async` to bind to the last call expr.
```
async a.b[3](1, 2).c[1](3, 4)
  |_______________________|
```

I'm not sure there is a way to express this with PEG, but if you have some good idea, I'm all ears.

**Comment by wqweto** [comment]
@Hejsil  You can try using non-consuming asserts (`&` and `!`) so your Option 1 from above can use a `SuffixOpBeforeLParen` rule like this:

    SuffixExpr
        <- AsyncPrefix (PrimaryExpr SuffixOpBeforeLParen?) FnCallArgumnets
         / (PrimaryExpr SuffixOp*)
         
    SuffixOpBeforeLParen
        <- SuffixOp &"(" !SuffixOp
         / SuffixOp SuffixOpBeforeLParen

Memoization in this PEG would be key for `SuffixOpBeforeLParen` performance not to suck though.

**Comment by Hejsil** [comment]
@wqweto I'm not 100% certain I understand your rule but doesn't it accomplish the same as:
```
SuffixExpr
    <- AsyncPrefix PrimaryExpr SuffixOp* FnCallArgumnets
     / PrimaryExpr (SuffixOp / FnCallArgumnets)*
```

Where, `SuffixOp` does not contain `FnCallArgumnets`

**Comment by wirelyre** [comment]
Since `SuffixOp` does not include function calls, why not:

```
SuffixExpr
    <- AsyncPrefix PrimaryExpr (SuffixOp* FnCallArguments+)+
     / PrimaryExpr (SuffixOp / FnCallArguments)*
```

or optionally expressly prohibit suffixes that are not function calls:

```
SuffixExpr
    <- AsyncPrefix PrimaryExpr (SuffixOp* FnCallArguments+)+ !SuffixOp
     / …
```

The second might be completely equivalent to the first; I haven't inspected the current grammar in detail.

**Comment by wqweto** [comment]
@Hejsil Opps, my bad! Just edited my post above with `!SuffixOpBeforeLParen` for the negative assert.

You can try [the sample PEG in this gist](https://gist.github.com/wqweto/44082b902efb7c45a6d990e467d72e1f) at https://pegjs.org/online with some sample input to figure out what gets parsed with the asserts implemented as above.

**Comment by andrewrk** [comment]
> Just another note here. [C# doesn't' use an async prefix to call async functions](https://docs.microsoft.com/en-us/dotnet/csharp/async). What do we gain from having this prefix, once coroutines change?

I think it's a reasonable idea to not require special syntax at the callsite for async functions. The only thing it would do is document the calling convention at the callsite. However, I am now convinced that isn't useful. Because in order to know the return type of the function call, you need to look at the function prototype, which has the calling convention as well as the return type.

So :+1: from me to deleting `async` at the callsite.

**Comment by Hejsil** [comment]
@andrewrk Alright, I'll leave this out of the grammar, but keep parsing of async calls until we rework coroutines.

**Comment by bbodi** [comment]
> > Just another note here. [C# doesn't' use an async prefix to call async functions](https://docs.microsoft.com/en-us/dotnet/csharp/async). What do we gain from having this prefix, once coroutines change?
> 
> I think it's a reasonable idea to not require special syntax at the callsite for async functions. The only thing it would do is document the calling convention at the callsite. However, I am now convinced that isn't useful. Because in order to know the return type of the function call, you need to look at the function prototype, which has the calling convention as well as the return type.
> 
> So 👍 from me to deleting `async` at the callsite.

I agree that this async is redundant information at the call site, but one other important goal of this keyword is to notify the reader that something unusual happens in that line.
Also it helps avoid accidents where the function signature was changed to async, but the call site is not aware of that and still assumes that the called (now async) function will be completed when the next line is executed.
I would like to be sure that ``bar()`` is called after ``foo()`` has been finished.
```
// No hidden control flow
var a = b + c.d;
foo();
bar();
```
I might be wrong, I don't fully know the current status of the async syntax, maybe that kind of dramatic change of the function signature is not even possible without changing the call site.

**Comment by andrewrk** [comment]
> something unusual happens in that line.

Is it unusual though? It's the same control flow. It's a function call, same as any other. The only difference is in the return type.

> Also it helps avoid accidents where the function signature was changed to async, but the call site is not aware of that and still assumes that the called (now async) function will be completed when the next line is executed.

This will always be caught by either:
 * The fact that the return type has changed from `T` to `(the coroutine frame type)`.
 * In the case of `void`, it will be caught by the compile error for ignoring the return value.

**Comment by andrewrk** [comment]
> What if call site uses type inference and does not use the return value for anything?

#335

**Comment by bbodi** [comment]
> #335

wow I like it, thanks for the link :)

**Comment by Hejsil** [comment]
Urg, just realized that all these are still problems as return exprs in my grammar:
```rust
fn a() if (true) A {}
fn a() while (true) A {}
fn a() for ([]void{}) A {}
fn a() comptime A {}
fn a() break A {}
fn a() cancel A {}
fn a() resume A {}
fn a() return A {}
```

**Comment by andrewrk** [comment]
I would be interested in a smaller scope project that only aimed to revert the #760 change with the `error` -> `anyerror` resolution (and didn't try to fix all the grammar). We can always come back to the formal grammar closer to 1.0.0. I see the syntax as a coat of paint - it's nice to know what color the end result will be, but we don't have to paint the corners and edges of things until the end.

If you have the motivation and energy to continue this project, by all means, carry on - it's high value important stuff - but if you're looking for an out, that could be it.

**Comment by Hejsil** [comment]
@andrewrk Naah, I'm still going strong. I'll probably just leave these constructs alone for now, and implement the grammar as it currently stands. It gets us 95% there

**Comment by Hejsil** [comment]
@UniqueID1 As far as I've come to understand from @andrewrk, the grammar on ziglang.org was meant to be something very similar to a PEG, just with some extra syntax to express some of the constructs more easily. The problem here is that the grammar was never tested so we had no ideas if the grammar actually fit how stage 1 parsed the language. This made it hard to implement the stage 2 parser, as the grammar and stage 1 parser contradicted in behavior.

With an actual PEG, we can test that our formal grammar actually can parse the language (using parser generators). This allows for syntax prototyping and makes it unambiguous how the Zig language should be parsed by a conforming compiler.

There is only one [syntactic rule](https://github.com/ziglang/zig/pull/1685#issuecomment-434651936) which I've not been able to specify in PEG. There are several options to solve this:
* Drop the grammar, implementation is king.
  * How do we distinguish intent from bugs?
  * How do we communicate the syntax in the language specification (#75) (I've heard that C++ have compilers that don't agree on what is valid syntax).
* Find another theoretical formal grammar that can handle these rules.
  * It has to be testable (parser generators) and unambiguous.
* Drop the rule. It adds no power to the language.
  * Violates the "only one way to do things"

**Comment by Hejsil** [comment]
@UniqueID1 The reason I went for PEG over CFG, was exacly because CFG was to limiting if unambiguaty was a goal, btw.

**Comment by Hejsil** [comment]
@UniqueID1 And the reason I wonna remove the `async` prefix, is because it's a complicated parsing rule, with not may actual benifits (if any). I wont the grammar to be simple, just like the language.

**Comment by Hejsil** [comment]
@UniqueID1 And as far as I know, most languages go for a CFG or PEG. Can you give an example of a language that does not?

**Comment by UniqueID1** [comment]
Nim uses extended BNF (CFG), but:

> The |, / symbols are used to mark alternatives and have the lowest precedence. / is the ordered choice that requires the parser to try the alternatives in the given order. / is often used to ensure the grammar is not ambiguous.

https://nim-lang.org/docs/manual.html

**Comment by Hejsil** [comment]
> I have all stage 1 tests passing locally. As I mentioned, we still have these being parsed incorrectly:
> 
> ```
> fn a() if (true) A {}
> fn a() while (true) A {}
> fn a() for ([]void{}) A {}
> fn a() comptime A {}
> fn a() break A {}
> fn a() cancel A {}
> fn a() resume A {}
> fn a() return A {}
> ```
> I can trivially make the `break`, `cancel`, `resume` and `return` cases work, as they can never return a `type`, so they can get a precedence above `TypeExpr`.
> 
> As for `comptime`, `if`, `for`, `while`, I think the plan is to have two syntaxes. One for an `if` type expression and one for `if` normal expression.

Implemented!

**Comment by wqweto** [comment]
Can confirm that I can parse all `.zig` sources in this PR (incl all tests) with [parser.y grammar](https://github.com/Hejsil/zig-grammar2/blob/master/peg/parser.y) as input to my (independent) [vbpeg parser generator](https://github.com/wqweto/VbPeg/tree/master/test/Runner/peg/Zig) which completely understands [Ian's peg/leg](http://piumarta.com/software/peg/) dialect for PEGs.

If time permits will annotate my fork of the grammar (in vbpeg tests) with native language actions to produce some kind of parse tree in JSON format for reference.

**Comment by Hejsil** [comment]
I ended up making the fmt passes, as I neede them myself:
* `zig-fmt-error-to-anyerror` `error -> anyerror`  
* `zig-fmt-revert-dot-init-container-decl` reverts #1628

**Comment by gernest** [comment]
@Hejsil  can you please help me understand how I can use `zig-fmt-error-to-anyerror` and `zig-fmt-revert-dot-init-container-decl`.

This change breaks a lot of stuff I have been working on, an automatic conversion will be helpful .
Thanks.

**Comment by Hejsil** [comment]
You will need to build the stage 2 compiler from these branches from source. Guide for that is in the read me

**Comment by gernest** [comment]
thanks
